File: _c++/symulacja_organizmow/Organizmy/Roslina.cpp
//
//  Roslina.cpp
//  WirtualnySwiat
//
//  Created by Michal Tarnacki on 10/04/2022.
//

#include "Roslina.hpp"

Roslina::Roslina(Swiat &swiat):Organizm(swiat){}
void Roslina::SetInicjatywa(){
    this->inicjatywa=0;
}
void Roslina::Akcja(){
    if(CzyMozeSieRozmnazac())
        Zasiej();
}
void Roslina::Zasiej(){
    int x = rand() % 100;
    if(x>=(this->GetPrawdopodobienstwo()))
        return;
    if(!SprawdzCzyMoznaBezpiecznieRuszyc(1))
        return;
    Polozenie nowePolozenie;
    bool uniq=true;
    while(true) {
        uniq=true;
        nowePolozenie.setXY(polozenie.GetX()+(rand()%3-1), polozenie.GetY()+rand()%3-1);
        if(nowePolozenie.GetX()>=0&&nowePolozenie.GetY()>=0&& nowePolozenie.GetX()<swiat.GetPlansza().GetN()&&nowePolozenie.GetY()<swiat.GetPlansza().GetM()){
            for (Organizm *organizm: swiat.GetOrganizmy()){
                if(nowePolozenie==organizm->GetPolozenie())
                    uniq=false;
            }
            if(uniq){
                break;
            }
        }
    }
    this->SetZdolnoscRozrodu(false);
    std::cout << "Roslina : " << this->Rysowanie() << " zasiala na polu: "<< nowePolozenie << std::endl;
    swiat.NarodzOrganizm(*this, nowePolozenie);
}
void Roslina::SetSila(){
    this->sila=0;
}

File: _java/symulacja_organizmow/src/Symulacja/Roslina.java
package Symulacja;

public abstract class Roslina extends Organizm{
    protected final void SetInicjatywa(){
        this.inicjatywa=0;
    };
    protected void SetSila(){

    };
    public Roslina(Swiat swiat){
        super(swiat);
    };
    public void Akcja(){
        if(CzyMozeSieRozmnazac())
            Zasiej();
    };
    public abstract char Rysowanie();
    public abstract int GetPrawdopodobienstwo();
    public final void Zasiej(){
        int x = (int)(Math.random()*100);
        if(x>=(this.GetPrawdopodobienstwo()))
            return;
        if(!SprawdzCzyMoznaBezpiecznieRuszyc(1))
            return;
        Polozenie nowePolozenie = new Polozenie();
        boolean uniq=true;
        int randx, randy;
        while(true) {
            randx=((int)(Math.random()*3)-1);
            randy=((int)(Math.random()*3)-1);
            uniq=true;
            nowePolozenie.Set(polozenie.GetX()+randx, polozenie.GetY()+randy);
            if(!(this.swiat instanceof SwiatHex)) {
                if (nowePolozenie.GetX() >= 0 && nowePolozenie.GetY() >= 0 && nowePolozenie.GetX() < swiat.GetPlansza().GetN() && nowePolozenie.GetY() < swiat.GetPlansza().GetM()) {
                    for (Organizm organizm : swiat.GetOrganizmy()) {
                        if (nowePolozenie.equals(organizm.GetPolozenie()))
                            uniq = false;
                    }
                    if (uniq) {
                        break;
                    }
                }
            }
            else{
                if(randx==-1&&randy==0)
                    nowePolozenie.Set(nowePolozenie.GetX()+2, nowePolozenie.GetY()-2);
                else if(randx==1&&randy==0)
                    nowePolozenie.Set(nowePolozenie.GetX()-2, nowePolozenie.GetY()+2);

                if (nowePolozenie.GetX() >= 0 && nowePolozenie.GetY() >= 0 && nowePolozenie.GetX() < swiat.GetPlansza().GetN() && nowePolozenie.GetY() < swiat.GetPlansza().GetM() && !(randx ==-1&&randy==-1) && !(randx ==1&&randy==1)) {
                    for (Organizm organizm : swiat.GetOrganizmy()) {
                        if (nowePolozenie.equals(organizm.GetPolozenie()))
                            uniq = false;
                    }
                    if (uniq) {
                        break;
                    }
                }

            }
        }
        this.SetZdolnoscRozrodu(false);
        swiat.zdarzenia.DodajString("Roslina : " + this.Rysowanie() + " zasiala na polu: " + nowePolozenie);
        swiat.NarodzOrganizm(this, nowePolozenie, TrybNarodz.Auto);
    };
    public abstract Organizm Urodz();
    public abstract boolean MozeBycPartnerem(Organizm organizm);
}

